### PR TITLE
N°7968 - ⚡️ Add faster way to rename column on the same table

### PR DESCRIPTION
**Pull request by Hipska**
## Base information
| Question                                                      | Answer 
|---------------------------------------------------------------|--------
| Related to a SourceForge thead / Another PR / Combodo ticket? | N/A
| Type of change?                                               | Enhancement

## Objective

The process of moving data from one column to another could be simplified a lot when both are on the same table.

## Proposed solution

Execute a `RENAME COLUMN` SQL statement instead of the following steps:
1. Create new column
2. Copy data from old to new column
3. Delete old column

## Checklist before requesting a review

- [x] I have performed a self-review of my code
- [x] I have tested all changes I made on an iTop instance
- [x] I have added a unit test, otherwise I have explained why I couldn't
- [x] Is the PR clear and detailed enough so anyone can understand digging in the code?

### CODE REVIEW

**Review by Molkobain — CHANGES_REQUESTED** [review]
Careful, this is compatible with MySQL 8.0+, but not MySQL 5.7 which is still [supported by iTop](https://www.itophub.io/wiki/page?id=3_2_0:install:requirements#itop_compatibility_list). You should stick to the `CHANGE` synthax:

| Database       | Version           | `CHANGE` Compatibility | `RENAME COLUMN` Compatibility |
|----------------|-------------------|-------------------------|-------------------------------|
| MySQL          | < 8.0             | Yes                     | [No](https://dev.mysql.com/doc/refman/5.7/en/alter-table.html)                            |
| MySQL          | 8.0+              | Yes                     | [Yes](https://dev.mysql.com/doc/refman/8.0/en/alter-table.html)                           |
| MariaDB        | < 10.5.2            | Yes                     | No                            |
| MariaDB        | 10.5.2+             | Yes                     | [Yes](https://mariadb.com/kb/en/alter-table/#rename-column)                            |

**Comment by Hipska** [comment]
Oh, that's probably the reason I didn't create the PR back then.

Downside with [CHANGE](https://dev.mysql.com/doc/refman/8.4/en/alter-table.html#alter-table-redefine-column) is that it needs the column definition, but there's a method for that, so fixed in 318a7e9.

**Review by Molkobain — APPROVED** [review]
`CMDBSource::GetfIeldSpec()` can return false, but I rather have an explicit error than hiding it. Good job.

**Comment by steffunky** [comment]
We already have some tests for this method in `tests/php-unit-tests/unitary-tests/setup/ModuleInstallerAPITest.php`, adding one testing this use case could be nice as @Molkobain already caught a regression in this method once
If you don't have time I'll try to add one sometime soon

**Comment by Hipska** [comment]
@Molkobain actually, the same method is doing the same without any checks: https://github.com/Combodo/iTop/blob/c6039f4b51fffacedc529df87ae7396c409e0c6e/setup/moduleinstaller.class.inc.php#L277
It will just never occur to return `false` because of the checks before.

**Review by steffunky — APPROVED** [review]
Works well on MariaDB 11.3.2 :+1:

**Comment by Hipska** [comment]
Funny, as that version isn't even [supported](https://www.itophub.io/wiki/page?id=latest:install:requirements#itop_compatibility_list) by Combodo.

File: tests/php-unit-tests/unitary-tests/setup/ModuleInstallerAPITest.php [hunk 1]
```diff
@@ -240,4 +240,48 @@ public function testMoveColumnInDB_MoveMultipleTable(): void
...
+			<<<SQL
+SELECT `{$sDstNonExistingColName}` FROM `{$sOrigTable}` WHERE `id` = 1
+LIMIT 1
+SQL
+		);
+
+		// Put data back in the original table
+		ModuleInstallerAPI::MoveColumnInDB($sOrigTable, $sDstNonExistingColName, $sOrigTable, $sOrigColName);
+
+		$this->assertEquals($sOrigValue, $sDstValue, "Data was not moved as expected");
```
Comment (Molkobain) [inline tests/php-unit-tests/unitary-tests/setup/ModuleInstallerAPITest.php:285]:
```suggestion
		// Try to move data
		ModuleInstallerAPI::MoveColumnInDB($sOrigTable, $sOrigColName, $sOrigTable, $sDstNonExistingColName);

		// Check if data was actually moved
		// - Either way, the column should exist
		$sDstValue = CMDBSource::QueryToScalar(
			<<<SQL
SELECT `{$sDstNonExistingColName}` FROM `{$sOrigTable}` WHERE `id` = 1
LIMIT 1
SQL
		);

		$this->assertEquals($sOrigValue, $sDstValue, "Data was not moved as expected");

		// Put data back in the original table
		ModuleInstallerAPI::MoveColumnInDB($sOrigTable, $sDstNonExistingColName, $sOrigTable, $sOrigColName);
```

I would rather put the assert there with your actual check. With the current proposition, it confused me, as I though you checked the data was put back in it's original column.

**Review by Molkobain — CHANGES_REQUESTED** [review]

File: tests/php-unit-tests/unitary-tests/setup/ModuleInstallerAPITest.php (hunk 1 above)
Comment (Molkobain) [reply tests/php-unit-tests/unitary-tests/setup/ModuleInstallerAPITest.php:285]: Otherwise the test is good, thanks!

Comment (Hipska) [reply tests/php-unit-tests/unitary-tests/setup/ModuleInstallerAPITest.php:285]: Doen't it break the state of the table when assert fails? (The way unit tests in php work are weird and confusing anyways)

Comment (steffunky) [reply tests/php-unit-tests/unitary-tests/setup/ModuleInstallerAPITest.php:285]: I wanted to keep consistency with the [code written by my ex Lead dev](https://github.com/Combodo/iTop/pull/676/files/007927c852879001fd30552b2ce6322b8371d911#diff-26ac6d70dfb1aa45415774deb778df56af8238e996d5d39f46cdc79cdef36e2dL137-L144)
Joking aside, wouldn't a failed assert stop the test execution and leave an inconsistent table?

Comment (Molkobain) [reply tests/php-unit-tests/unitary-tests/setup/ModuleInstallerAPITest.php:285]: 🙈🙈🙈

I think both your statements are valid, I therefore withdraw mine 😁

**Review by Molkobain — APPROVED** [review]

**Comment by rquetiez** [comment]
To take all the benefits out of this proposal, the table alteration algorithm must be at least `inplace`.

I've just conducted a quick study to check which DDL algorithm is used, and make sure that the syntax `ALTER TABLE table CHANGE column...` would not imply the infamous `COPY` algorithm.

### Conclusion
The proposed implementation will be quick in any versions supported by iTop, excepted for the rare cases when the `COPY` algorithm is still used (e.g. table created with mariaDB < 10.3 and never rebuilt since the upgrade of the server).

### Methodology
As an alternative to measuring the time spent on executing the `alter table` statement on a "huge" table, one can specify `algorithm=inplace`, so that the `alter table` statement would fail if the engine is not able to execute the `INPLACE` algorithm (or a quicker algorithm).

I've also tested with `lock=none` to document wether the operation involves locking or not (See [Online DDL Performance and Concurrency](https://dev.mysql.com/doc/refman/8.4/en/innodb-online-ddl-performance.html))

```mysql
alter table organization change code new_code varchar(255) default '' null, algorithm=inplace, lock=none;
alter table organization change code new_code varchar(255) default '' null, algorithm=instant;
```

### Results
| Database       | Version           | algorithm=`inplace`, lock=none | algorithm=`instant` |
|----------------|-------------------|-------------------------|-------------------------------|
| MySQL          | 5.7               | Yes                     | n/a |
| MySQL          | 8.3               | Yes                     | Yes |
| MariaDB        | 10.3              | Yes                     | Yes |

**Review by rquetiez — APPROVED** [review]

**Comment by jf-cbd** [comment]
Thanks a lot @Hipska for the contribution !
